src/audio/detect_test.c

```diff
@@ -0,0 +1,484 @@
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int test_keyword_cmd(struct comp_dev *dev, int cmd, void *data,
+			    int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+
+	trace_keyword("test_keyword_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		return test_keyword_ctrl_set_data(dev, cdata);
+	case COMP_CMD_GET_DATA:
+		return test_keyword_ctrl_get_data(dev, cdata, max_data_size);
+	default:
+		return -EINVAL;
+	}
+}
+
+static int test_keyword_trigger(struct comp_dev *dev, int cmd)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_keyword("test_keyword_trigger()");
+
+	ret = comp_set_state(dev, cmd);
+	if (ret)
+		return ret;
+
+	switch (cmd) {
+	case COMP_TRIGGER_START:
+	case COMP_TRIGGER_RELEASE:
+		cd->detected = 0;
+		break;
+	}
+
+	return ret;
+}
+
+/*  process stream data from source buffer */
+static int test_keyword_copy(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+
+	tracev_keyword("test_keyword_copy()");
+
+	/* keyword components will only ever have 1 source */
+	source = list_first_item(&dev->bsource_list,
+				 struct comp_buffer, sink_list);
+
+	/* make sure source component buffer has enough data available for copy
+	 * Also check for XRUNs
+	 */
+	if (source->avail < cd->period_bytes) {
+		trace_keyword_error("test_keyword_copy() error: "
+				   "source component buffer"
+				   " has not enough data available");
+		comp_underrun(dev, source, cd->period_bytes, 0);
+		return -EIO;	/* xrun */
+	}
+
+	/* copy and perform detection */
+	cd->detect_func(dev, source, dev->frames);
+
+	/* calc new available */
+	comp_update_buffer_consume(source, cd->period_bytes);
+
+	return dev->frames;
+}
+
+static int test_keyword_reset(struct comp_dev *dev)
+{
+	trace_keyword("test_keyword_reset()");
+
+	return comp_set_state(dev, COMP_TRIGGER_RESET);
+}
+
+static int test_keyword_prepare(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int ret;
+
+	trace_keyword("test_keyword_prepare()");
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret)
+		return ret;
+
+	if (cd->config) {
+		ret = alloc_mem_load(cd, cd->config->load_memory_size);
+		if (ret < 0)
+			goto err;
+	}
+
+	return 0;
+
+err:
+	comp_set_state(dev, COMP_TRIGGER_RESET);
+	return ret;
+}
+
+static void test_keyword_cache(struct comp_dev *dev, int cmd)
+{
+	struct comp_data *cd;
+
+	switch (cmd) {
+	case CACHE_WRITEBACK_INV:
+		trace_keyword("test_keyword_cache(), CACHE_WRITEBACK_INV");
+
+		cd = comp_get_drvdata(dev);
+
+		dcache_writeback_invalidate_region(cd, sizeof(*cd));
+		dcache_writeback_invalidate_region(dev, sizeof(*dev));
+		break;
+
+	case CACHE_INVALIDATE:
+		trace_keyword("test_keyword_cache(), CACHE_INVALIDATE");
+
+		dcache_invalidate_region(dev, sizeof(*dev));
+
+		cd = comp_get_drvdata(dev);
+		dcache_invalidate_region(cd, sizeof(*cd));
+		break;
+	}
+}
+
+struct comp_driver comp_keyword = {
```

Comment:
Please see c06a65a8d0915856f49ceb8241af8540d79c4b44 and update to self register.

Reply:
Would you mind elaborating? I thought I already updated it